Add ProxySSH implementation to transportv1.Service

rosstimothy opened the pull request:
Completes the service implementation of the transportv1.Service. Ensures that both SSH and SSH Agent protocols are multiplexed correctly over the gRPC stream via `TestService_ProxySSH`.

The access checker still needs to be populated when dialing, however, the authentication verification and session control pieces are not yet in place. Once they get added we will be able to extract the user identity and provide the access checker to the dialer.

Part of #19812

EdwardDowling commented on lib/srv/transport/transportv1/transport.go:
```diff
@@ -125,3 +140,124 @@ func (c clusterStream) Recv() ([]byte, error) {
 	return trace.Wrap(c.stream.Send(&proxyv1.ProxyClusterResponse{Frame: &proxyv1.Frame{Payload: frame}}))
 }
+
+func (s *Service) ProxySSH(stream proxyv1.ProxyService_ProxySSHServer) error {
```
Could you add some comments on the exported function in this file?

fspmarshall commented on lib/srv/transport/transportv1/transport.go:
```diff
@@ -125,3 +149,128 @@ func (c clusterStream) Recv() ([]byte, error) {
+				return
+			}
+
+			switch frame := req.Frame.(type) {
+			case *proxyv1.ProxySSHRequest_Ssh:
+				sshStream.incomingC <- frame.Ssh.Payload
+			case *proxyv1.ProxySSHRequest_Agent:
+				agentStream.incomingC <- frame.Agent.Payload
+			default:
+				s.cfg.Logger.Errorf("received unexpected ssh frame: %T", frame)
+				continue
+			}
```
Nit: I'm guessing these sends don't select on `stream.Context().Done()` because we want to ensure that all received messages are consumed, even if the stream is closing... assuming that's the case, maybe worth adding comments to `sshStream.Recv` indicating that failure to drain could result in a leak?

fspmarshall commented on lib/srv/transport/transportv1/transport.go:
```diff
@@ -125,3 +149,128 @@ func (c clusterStream) Recv() ([]byte, error) {
+		return trace.BadParameter("unable to find peer")
+	}
+
+	// TODO(tross): add access checker once authentication pieces are in place
+	conn, err := s.cfg.Dialer.DialHost(ctx, p.Addr, host, port, req.DialTarget.Cluster, nil, s.cfg.agentGetterFn(agentStreamRW))
+	if err != nil {
+		return trace.Wrap(err, "failed to dial target host")
+	}
```
Can we add something to break this outside of tests until access checker is in place?  Maybe an add an `insecureUnauthenticated` flag and have this panic if we get here without that flag enabled?  Bit silly, but I hate the idea of merging insecure-by-default logic to master, even if it doesn't currently do anything.  Never know when a bad merge or human error might make it pop up unexpectedly.

fspmarshall reviewed (approved):
LGTM w/ minor suggestions.

rosstimothy committed:
Add ProxySSH implementation to transportv1.Service

Completes the service implementation of the transportv1.Service.
Ensures that both SSH and SSH Agent protocols are multiplexed
correctly over the gRPC stream via TestService_ProxySSH.

Part of #19812
